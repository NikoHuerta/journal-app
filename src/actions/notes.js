import { db } from "../firebase/firebaseConfig";
import { collection, addDoc, updateDoc, doc , deleteDoc} from "firebase/firestore";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";


export const startNewNote = () => { //tarea async
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        try{
            const doc = await addDoc(collection(db, `${uid}/journal/notes`), newNote);
            dispatch(activeNote(doc.id, newNote));
            dispatch(addNewNote(doc.id, newNote));
        }catch(e){
            console.log(e);
        }
    }
}

export const startLoadingNotes = (uid) => { //tarea async
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}

export const addNewNote = (id, note) => ({
    type: types.notesAddNewNote,
    payload: { 
            id,
            ...note
    }
});

export const activeNote = (id, note) => ({ //tarea sync
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const setNotes = (notes) => ({ //tarea sync (local)
    type: types.notesLoad,
    payload: notes
});

export const startSaveNote = ( note ) => { //tarea async (remota)
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        // //obtener el estado actual de las notas desde state redux
        // const { notes:notesReducer } = getState().notes;

        if(!note.url){
            delete note.url;
        }

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        const noteRef = doc(db, `${ uid }/journal/notes/${ note.id }`);
        try {
                await updateDoc(noteRef, noteToFirestore);
                dispatch(refreshNote(note.id, noteToFirestore));

                // //aÃ±adir note nueva al state
                // notesReducer.push(note);
                // dispatch(setNotes(notesReducer));

                Swal.fire('Saved', note.title, 'success');
        }catch(e){
            Swal.fire('Error Saving', e, 'error');
        }
    }
};

export const refreshNote = (id, note) => ({ //tarea sync (local)
    type: types.notesUpdated,
    payload: { 
        id,
        note: {
            id,
            ...note
        }
    }   
});

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active:activeNote } = getState().notes;

        Swal.fire({
                title: 'Uploading...', 
                text: 'Please wait...',
                allowOutsideClick: false,
                didOpen: () => {
                    Swal.showLoading();
                }
            });

        const fileUrl = await fileUpload(file);
        Swal.close();

        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote));
        // console.log(fileUrl);
    }
};

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        const noteRef = doc(db, `${ uid }/journal/notes/${ id }`);

        try {
            await deleteDoc(noteRef);
            dispatch(deleteNote(id));

        } catch (error) {
            console.log(error);
        }
    }
};

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const logoutNote = () => ({
    type: types.notesLogoutCleaning
})