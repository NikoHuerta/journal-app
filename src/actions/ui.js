import { types } from "../types/types";

export const setError = (error) => ({
    type: types.uiSetError,
    payload: error
});

export const removeError = () => ({
    type: types.uiRemoveError
});