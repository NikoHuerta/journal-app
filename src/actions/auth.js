import Swal from 'sweetalert2';

import { getAuth, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from 'firebase/auth';
import { googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/types"
import { finishLoading, startLoading } from './ui';
import { logoutNote } from './notes';


export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        
        dispatch(startLoading());

        const auth = getAuth();
        // console.log(auth);

        return signInWithEmailAndPassword(auth, email, password)
            .then( ({user}) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
            })
            .catch(e => {
                dispatch(finishLoading());
                //console.log(e);
                let msg_err;
                ((e.code ==='auth/user-not-found') || (e.code ==='auth/wrong-password')) ? msg_err= 'You have entered an invalid username or password, or your account does not exist' : msg_err = 'Service error';
                Swal.fire('Error', msg_err, 'error');

            });
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        const auth = getAuth();
        return createUserWithEmailAndPassword(auth, email, password)
            .then( async ({ user }) => {

                await updateProfile(user, { displayName: name });                
                dispatch(login(user.uid, user.displayName));
            })
            .catch(e => {
                
                let msg_err;
                (e.code ==='auth/email-already-in-use') ? msg_err= 'The email address is already in use by another account.' : msg_err = 'Service internal error.';
                Swal.fire('Error', msg_err, 'error');
                return msg_err;
            });

    }
}

export const startGoogleLogin = () => {

    return(dispatch) => {
        dispatch(startLoading());

        const auth = getAuth();
        return signInWithPopup(auth, googleAuthProvider)
            .then( ({ user }) => {
                dispatch(
                    login(user.uid, user.displayName)
                );
                dispatch(finishLoading());
            })
            .catch(e => {
                Swal.fire('Error', 'The Google Login PopUp was closed.', 'error');
                dispatch(finishLoading());
                return e.code;
            });
    }
}

export const startLogout = () => {
    return (dispatch) => {
        const auth = getAuth();
        //console.log(auth);
        return signOut(auth)
            .then( ( ) => {
                dispatch(logout());
                dispatch(logoutNote());
            })
            .catch( e => {
                console.log(e);
            });
    }
}

export const login = (uid, displayName) => ({
        type: types.login,
        payload: {
            uid,
            displayName
        }
});

export const logout = () => ({
    type: types.logout
});


