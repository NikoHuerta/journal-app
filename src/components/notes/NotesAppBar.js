import moment from 'moment';
// import 'moment/locale/es';

import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/notes'

export const NotesAppBar = () => {

    const dispatch = useDispatch();
    const { active } = useSelector(state => state.notes);
    
    // moment.locale('es');
    const noteDate = moment(active.date);
    

    const handleSave = () => {
        dispatch(startSaveNote(active));
    }

    const handlePictureClick = () => {
        // console.log('picture clicked');
        document.querySelector("#fileSelector").click();
    }

    const handleFileChange = (e) => {
        //console.log(e);
        const file = e.target.files[0];
        //console.log(file);
        if (file){
            dispatch(startUploading(file));
        }
    }

    return (
        <div className="notes__appbar">
            <span>{ noteDate.format('LLLL') }</span>

            <input
                type="file"
                style={{ display: 'none' }}
                onChange={ handleFileChange }
                id="fileSelector"
            >
            </input>

            <div>
                <button 
                    className="btn"
                    onClick={ handlePictureClick }
                >
                    Picture
                </button>

                <button 
                    className="btn"
                    onClick={ handleSave }
                >
                    Save
                </button>

            </div>
        </div>
    )
}
