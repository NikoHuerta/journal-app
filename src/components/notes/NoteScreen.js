import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const { active:note } = useSelector(state => state.notes);
    
    const [formValues, handleInputChange, reset] = useForm( note );
    const { title, body } = formValues;
    
    const activeId = useRef(note.id);

    useEffect(() => {
        
        if(note.id !== activeId.current){
            reset(note);
            activeId.current = note.id;
        }
        
    }, [note, reset]);

    useEffect(() => {

        dispatch(activeNote(formValues.id, {
            ...formValues
        }));
        
    }, [formValues, dispatch]);

    const handleDelete = () => {
        //console.log('Delete Clicked');
        dispatch(startDeleting(note.id));
    }


    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">
                <input 
                    type="text"
                    placeholder="Title"
                    className="notes__title-input"
                    autoComplete='off'
                    name='title'
                    
                    onChange={ handleInputChange }
                    value={ title }
                />

                <textarea
                    placeholder="Description"
                    className="notes__textarea"
                    autoComplete='off'
                    name='body'
                    
                    onChange={ handleInputChange }
                    value={ body }
                ></textarea>

                {
                    (note.url)&&
                    (<div className="notes__image">
                        <img
                            src={ note.url }
                            // onChange={ handleInputChange }
                            alt="imagen"
                            name='url'
                        />

                    </div>)
                }
            </div>
            
            <button 
                className='btn btn-danger'
                onClick={ handleDelete }
            >
                Delete
            </button>

        </div>
    )
}
