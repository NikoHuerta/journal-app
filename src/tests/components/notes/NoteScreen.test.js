import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import '@testing-library/jest-dom';

import { NoteScreen } from "../../../components/notes/NoteScreen";
import { MemoryRouter } from "react-router-dom";
import { activeNote } from "../../../actions/notes";

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {
        uid: '955fp6FcxWfqIKnsJj1EKpXeT1T2',
        name: 'Nicolas Testing'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes:{ 
        notes: [],
        active: {
            id: 1234,
            title: 'hola',
            body: 'mundo',
            date: 0
        }
    }
};
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(<Provider store={ store }>
                        <NoteScreen />
                    </Provider>);

describe('Pruebas sobre <NoteScreen />', () => {
   
    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de disparar la accion activeNote', () => {
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Hola titulo actualizado'
            }
        });
        expect(activeNote).toHaveBeenLastCalledWith(
            initState.notes.active.id,
            {
                id: initState.notes.active.id,
                title: 'Hola titulo actualizado',
                body: initState.notes.active.body,
                date: initState.notes.active.date
            }
        );

    });


    
    
});
