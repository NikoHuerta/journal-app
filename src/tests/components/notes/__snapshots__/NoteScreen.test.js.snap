// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <NoteScreen /> Debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      Wednesday, December 31, 1969 9:00 PM
    </span>
    <input
      id="fileSelector"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Title"
      type="text"
      value="hola"
    />
    <textarea
      autoComplete="off"
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="Description"
      value="mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
